Zenith Checks

sipsma opened the pull request:
This PR is the first chunk being split out from the original [Zenith PR](https://github.com/dagger/dagger/pull/5443) where we did prototype implementations of all the different entrypoints.
* Splitting that monolith PR out into chunks is intended to make API bikeshedding more manageable and give us an opportunity to cleanup code from a fresh base, which should be much faster than trying to update the original PR.
* Anyone working on Zenith should feel free to push to this PR. I'll also cherry-pick as much as feasible from the original PR.

It's intended to consist of:
1. Base `Environment` API + implementation
2. `Check` APIs + implementation
3. Relevant CLI commands (`dagger env ...`, `dagger checks`)
4. Support for the above in Go + Python SDKs
5. Test coverage for all of the above.
6. The ability to run the above tests as checks in the dagger CI env :-)

sipsma commented on core/schema/env.graphqls:
```diff
@@ -0,0 +1,98 @@
+scalar CheckResultID
+
+"A group of Dagger entrypoints that can be queried and/or invoked."
+type Environment {
```
Our longer term idea is for this to become an `interface` instead. I didn't include that here yet because I want to reserve the right to save that for follow up PRs.

The reason being that I *suspect* adding support for graphql interfaces may end up being a very significant chunk of work since it requires updates to the codegen for every SDK. That work will end up being worth it (I imagine we will have plenty of other use cases besides this one), but I don't necessarily want this PR to get blocked on that effort.

TomChv replied:
This would be pretty tricky to implement though but I can def see the benefit of supporting interface

sipsma replied:
I think the rough outline would be that when there's a gql `interface`, the codegen produces whatever is closest to that concept in it's language:
1. For Go and Typescript, it's literally just an `interface` in those language's syntax (ts has those iirc, right?)
2. For Python, I guess it might literally just be a type annotation since the underlying objects are all dynamic anyways. Not sure though
3. For Rust, a trait? Again, not sure
4. etc.

I think that actually may be all that's fundamentally needed.

It would at least be nice-to-have the ability to also convert the interface into a concrete object (possible in [raw graphql](https://www.apollographql.com/docs/apollo-server/schema/unions-interfaces/#querying-an-interface), in go it would be `foo, ok := ifaceObject.(Foo)`, etc.). But I'm not sure if there is actually a hard requirement for that, so hopefully could be left out for sake of simplicity.

So honestly it _might_ not be too much work for an individual SDK, at least in this early "theoretical" stage. I'm just more concerned that it's `work * numSDKs` :-)

TomChv replied:
> For Go and Typescript, it's literally just an interface in those language's syntax (ts has those iirc, right?)

For typescript it would be an `abstract class` or an interface, depending on what we want at the end.

> For Rust, a trait? Again, not sure

Yep, a `trait` shall do the job

> So honestly it might not be too much work for an individual SDK, at least in this early "theoretical" stage. I'm just more concerned that it's work * numSDKs :-)

Yeah totally but once it's done, we shall not touch it anymore, I think it's worth the time spent, I'm just scared that it change API at some point, I might be wrong tho. 

I do not have a clear view of how API shall look using interface

sipsma commented on core/schema/env.graphqls:
```diff
@@ -0,0 +1,98 @@
+}
+
+"The result of an environment's check."
+type CheckResult {
```
I am leaving out the `Subresult` apis in the original Zenith PR from here for now. I agree with the previous sentiments that it would be better if we can leave those out, so starting without them unless a strong need is proven.

sipsma commented on core/schema/env.graphqls:
```diff
@@ -0,0 +1,98 @@
+  "This check with the given subcheck"
+  withSubcheck(id: CheckID!): Check!
+
+  """
+  This check with the given container used to determine the check's result.
+  If set, the container will be executed and the check result will be set to
+  success if the container exits with a zero exit code, failure otherwise.
+  """
+  withContainer(id: ContainerID!): Check!
```
This API was suggested by @vito while he was implementing dynamic checks derived from go tests: https://github.com/dagger/dagger/blob/2763c2241e88a3282622652ff7309550982b473f/universe/go/main.go#L179

It's extremely convenient for those sorts of dynamic use cases, so I think we should start with it in the API. If we can find a better way of supporting dynamic checks that only relies on entrypoint env code, then it may become slightly redundant and we can consider whether we still need it.

sipsma commented on core/schema/env.graphqls:
```diff
@@ -0,0 +1,98 @@
+}
+
+"An entrypoint for tests, lints or anything that can pass/fail."
+type Check {
```
I also left out the previous APIs for check flags (i.e. parameters to checks) for now given our most recent conversation on the topic. This is another case where I think it's worth starting here and seeing how far we can get without them before considering re-adding.

sipsma commented on core/schema/env.graphqls:
```diff
@@ -0,0 +1,110 @@
+  success: Boolean!
+
+  "This check result with success set to the given value."
+  withSuccess(success: Boolean!): CheckResult!
+
+  "Any output associated with this check result."
+  output: String!
+
+  "This check result with the given output."
+  withOutput(output: String!): CheckResult!
```
In my in-progress reimplementation of checks, I currently am going to get rid of `withSuccess` and `withOutput` here and instead just have a field for constructing results from static success+output values.

Should fix the problem where not calling `.WithSuccess(true)` creates an implicit fail (which became confusing to explain during demos) and the fact that these methods are technically callable by anyone despite the fact that they should really be set at the point the result is evaluated and mutable after.

Will remove here once I've confirmed this theory works in practice.

TomChv reviewed (commented):
Left minimal comments as first review, I might fix these comments myself

sipsma commented:
Pushed an update w/ basically working refactored implementation.

Very happy with how things have cleaned up so far. A lot of the convoluted logic is just gone now via refactoring to have a clean split between `core/schema` doing all graphql stuff and `core` all the environment specific logic.

Subtleties around handling multiple possible return types for checks, but it's now all concentrated into one block of code rather than splattered all over the place.

By far the biggest time sink was that I realized the previous implementation of `currentEnvironment` was mostly broken, only worked under extremely specific scenarios. That's fixed now.

What makes that API extremely tricky is that to be fully correct+consistent, you need to have it return the exact environment w/ ID matching the one currently being executed in, but that creates circular dependencies. After many attempts, I found a solution using `CacheMap` + our trick for passing uncached data to execs that seems to work so far. It's by far the most subtle behavior, but thankfully it should be totally re-usable for the rest of the entrypoint types.

sipsma commented:
@vito just pushed a first attempt at the client-side display of checks; it looks okay-ish but I couldn't get the hierarchy to display exact how I want.
<img width="895" alt="Screenshot 2023-08-21 at 8 26 14 PM" src="https://github.com/dagger/dagger/assets/30126853/fa950c63-c967-4171-a56d-2ff30996d619">

E.g. there it's showing `sad-composite-check-from-dynamic-dep->yet-another-sad-composite-check` but I really just wanted it to show `yet-another-sad-composite-check`. But when I tried to change it I ended up not getting any indentation at all. 

Haven't totally groked the api for all that yet, so let me know if I'm just missing something obvious.

That screenshot is from running `dagger checks` in `core/integration/testdata/environments/go/basic` if you want to try it too.

TomChv commented on core/pythonruntime.go:
Done in https://github.com/dagger/dagger/pull/5659/commits/5122e3fa395851e916a59a8062c4c0b4ea6b9482

vito commented:
@sipsma I'll see if I can figure it out! For what it's worth it took a few attempts on the `zenith` branch, and here's what I finally landed on:

https://github.com/sipsma/dagger/blob/d3973bf62288fab0dd3176a70aa97d53f0ac11e0/core/schema/env.go#L683-L713

In a nutshell:

1. Create a group for each environment, so that multi-env check hierarchies are reflected. This also adds the outermost group in our demo, since there's the toplevel environment.
2. If the current check is a leaf node, initialize a vertex for it and report the check results.
3. Initialize _another_ group for each vertex, _after_ initializing the leaf node vertex, so that it's not a member of its own group. This extra group is for any Checks returned by the vertex. It's important that the leaf node's vertex itself isn't a member of this group, because that just results in extra noise. If nothing gets placed in this group, it will just not be displayed, so it doesn't hurt to create it before knowing whether we'll need it.

vito commented on core/env.go:
```diff
@@ -0,0 +1,655 @@
+	if err := resourceid.Decode(&check, id); err != nil {
+		return nil, err
+	}
+	if !check.IsCheck {
```
Could be worth cherry-picking this if you agree with the idea: https://github.com/sipsma/dagger/commit/45fc541290ae0d3609a5c6d3f82718f805e222f7

It changes IDs to look like `core.ContainerID:eyJ...` and verifies the type matches the expected type when decoding. It's a little janky that it's tightly coupled to Go's `%T` representation, so that could be tweaked, but this is all internal plumbing at the moment anyway.

Tangentially related, I came across [RFC-6920](https://www.rfc-editor.org/rfc/rfc6920) today (via [this post](https://posts.summerti.me/rfc-6920/)) which seems like a reasonable candidate format for our IDs. Maybe something like this?:

```
ni://dagger.io/sha-256;f4OxZX_x_FO5LcGBSKHWXfwtSx-j1ncoSt3SABJtkGk?ct=application/x-dagger-container
```

Anyway, that's neither here nor there, just cool to see a relevant RFC, with some kind of fun implications like having `https://dagger.io/.well-known/ni/sha256/f4OxZX_x_FO5LcGBSKHWXfwtSx-j1ncoSt3SABJtkGk` do... something interesting.

sipsma replied:
Oh nice thank you! I have a `TODO:` comment somewhere about backporting the cleaner fix to all this but hadn't taken the time to go track down which commit it was, appreciate it!

> Tangentially related, I came across [RFC-6920](https://www.rfc-editor.org/rfc/rfc6920) today (via [this post](https://posts.summerti.me/rfc-6920/)) which seems like a reasonable candidate format for our IDs. Maybe something like this?:

Ah neat! Honestly I'm still holding out that someday we'll have time to finally do the work needed to persist hash<->id mapping in the cache and be rid of needing to deal with anything like this, *but* the URL approach is super intriguing nonetheless since it would still probably make sense to support something akin to the current IDs externally as a "serialized DAG" format (similar to what you did w/ Bass IIRC)

TomChv replied:
> Tangentially related, I came across [RFC-6920](https://www.rfc-editor.org/rfc/rfc6920) today (via [this post](https://posts.summerti.me/rfc-6920/)) which seems like a reasonable candidate format for our IDs. Maybe something like this?:

I love that too

Quick question tho about the article:

> And while [CDNs do support caching subsets of a whole object](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/RangeGETs.html), there is no easy way for a RFC 6920 client to check whether their partial content is valid or not, as it uses a hash over the whole data, and not over a tree of parts.

Can this create issue with our cache? Like valid cache that should be invalidated or vice-versa? I do not have the full context of the id work but I'm trying to understand the whole logic.

vito replied:
> Can this create issue with our cache? Like valid cache that should be invalidated or vice-versa? I do not have the full context of the id work but I'm trying to understand the whole logic.

@TomChv I don't think we would need to care about partial fetches/etc, since the object being referenced is relatively small - essentially a recipe for building a container image, and not the container image itself.

sipsma commented:
Thank you @vito for the improvements to the check output, looks much better now!

<img width="905" alt="Screenshot_2023-08-22_at_6 49 06_PM" src="https://github.com/dagger/dagger/assets/30126853/85a24b8e-966c-4096-84b6-83f7f8e3314f">

---

Just pushed an update to the `ci/` env to delete the old project command code and now have a check for the engine integ tests, it's (mostly) working!
* There are random test failures here and there, various things related to dagger-in-dagger (networking stuff mostly from what I can tell). Will take a closer looks but nothing that's really related to this PR deeply as far as I can tell so far.
* One meaningful problem I ran into is that I implemented the check using `dag.Check().WithContainer(...)`, but the container I returned had a local dir embedded in it. This caused problems because the `Container` is only evaluated later after the env has exited.
  * Worked around atm by just putting a `.Sync` call in the env code. There's various options for a "real fix"; my current favorite would be to take advantage of the fact that env entrypoints are actually just `Container.WithExec` calls and map any `Host.Directory` calls embedded in the LLB they return to read the files/dirs from the output of that Exec. I'm hoping `opDAG` can make that *relatively* straightforward
  * Will save that for a follow-up though since it's not actually related to checks per-se, independent problem that will impact entrypoints in general.

TomChv commented:
> various things related to dagger-in-dagger (networking stuff mostly from what I can tell). Will take a closer looks but nothing that's really related to this PR deeply as far as I can tell so far.

We have that kind of issues in all our CI PR, all execution in `dagger-runner` are failing due to network errors (only few tests).

Do you think that it will also solve theses issues?

sipsma commented:
> > various things related to dagger-in-dagger (networking stuff mostly from what I can tell). Will take a closer looks but nothing that's really related to this PR deeply as far as I can tell so far.
> 
> We have that kind of issues in all our CI PR, all execution in `dagger-runner` are failing due to network errors (only few tests).
> 
> Do you think that it will also solve theses issues?

I sort of doubt it (would be happy to be wrong), the error I'm getting a lot in the new check is stuff like `buildkitd: plugin type="bridge" failed (add): failed to create bridge "remote-cache-0env0": could not add "remote-cache-0env0": numerical result out of range`, which I haven't seen anywhere else before.

Haven't tried debugging yet but I'm guessing I am just misconfiguring networking stuff or somehow hitting some new weird corner case due to the fact that this dagger-in-dagger setup is subtly different than our previous one (we are setting up the dagger-in-dagger engine from an environment, which is a nested exec that has API access back to the engine its running it, rather than being a "host client" of the engine)

vito commented on core/schema/env.graphqls:
```diff
@@ -0,0 +1,103 @@
+  success: Boolean!
+
+  "Any output obtained from evaluating the check's success."
+  output: String!
```
WDYT of adding a duration field here? I think that'd be important for measuring individual test performance over time. (And overall suite performance too.)

sipsma replied:
Oh I like that a lot! I think we'd probably want it to be measured by the shim and then persisted next to `stdout/stderr/exitCode` on the exec meta mount so that it's cached and linked to the exec itself.

Probably also worth storing it as `startTime/stopTime` unix timestamps so that we have the ability distinguish between "sum of durations of all checks" and "wall clock time" when stuff is running in parallel (maybe... some trickiness there obviously).

I don't *think* that's probably very difficult at all? So I'll give it a shot, but otherwise I'll just save for a followup.

sipsma commented:
Ran into some really tricky corner cases around environments and caching after adding the `EngineTests` check to the `ci/` env. Fortunately found some pretty nice solutions that should squash those problems for both `Check` and the rest of our entrypoint types; they also improve the safety around how we handle local dirs in general for all of Dagger too.

The problems arose due to the fact that:
1. The execution of environment entrypoint code and the results it returns are cached
   * This is a very good thing for performance and functionality, not a problem in and of itself.
2. The results returned by environment entrypoint code can have various pieces of ephemeral/session-specific data embedded in them, e.g.
   * Local dir references that point to data created in the entrypoint exec (i.e. the entrypoint code creates some file and then loads it into dagger via `dag.Host().Directory()`)
   * Service dependencies, which currently tie into session data due to DNS names being based on caller client IDs
   * Secrets created via `SetSecret`, which stores the secret in memory for the session

As a result, it was possible for us to load a cached entrypoint result, evaluate it, but then get errors because it refers to stuff that no longer exists (it was from a previous session where the entrypoint result was uncached and thus evaluated).

The solution comes in a few parts:
1. Local dir imports are now implemented using a new custom source type called `blob://` that lets us refer to the import as its content-addressed ID rather than as LLB instructions for loading a particular path from a particular client.
   * The implementation still does the same local dir sync as before, but rather than returning LLB for the `local://` source, we put the synced dir into the content store and then create `blob://` source LLB that points to that instead.
   * The part that requires extra careful handling is that buildkit would still be free to prune `blob://` cache refs without pruning the entrypoint exec cache ref.
   * To solve that, after an entrypoint is evaluated, we walk the DAG of its result and gather all the `blob://` contents it has a dependency on. Then we use the `LeaseManager` to add a dependency between the entrypoint exec cache ref and all those blobs, which ensures the blobs can't be pruned while the exec cache ref still exists.
2. The problem w/ service DNS names is fixed by just re-arranging where the dns metadata gets provided. It used to be in `Container.WithExec`, which resulted in that metadata being embedded in the `ContainerID`, but now its set only when an Exec is actually being evaluated.
   * This is done in `buildkit.Client.Solve` via `OpDAG` rewriting
3. The problem w/ `SetSecret` still exists, but the idea will be to just not allow `SetSecret` to be called in env code.
   * I don't believe there's any use case for this. Instead, secrets should always come from "the outside".
   * For now, "the outside" could only be client callers (where it would presumably be passed as a parameter one way or another)
   * Once we add support for `SecretProvider` as an entrypoint type, that will be another valid source. However, that implementation will not use `SetSecret`, instead it will be invoked "just-in-time" when a secret is needed (and without caching)

---

With all that in place, the errors I was getting when trying to run the `EngineTests` check after it has been cached are all gone; it runs fully cached now!

And like I mentioned, this solution will also squash the exact same problems for `Artifacts` and the other entrypoint types, so we *should* be on much more solid footing going forward now.

vito reviewed (commented):
Submitting partial review early for max async since I left a giant comment, I think I'll just go back to single comments instead of batched reviews, apologies in advance for notification spam

vito commented on cmd/dagger/check.go:
```diff
@@ -0,0 +1,227 @@
+func RunCheck(ctx context.Context, _ *client.Client, c *dagger.Client, env *dagger.Environment, cmd *cobra.Command, dynamicCmdArgs []string) (err error) {
+	rec := progrock.RecorderFromContext(ctx)
+
+	// TODO(vito): this is pretty confusing, but we need to initialize a root
+	// group for subsequent groups + vertices
+	rec = rec.WithGroup(progrock.RootGroup)
```
This can be removed after #5681 is merged (details: ["Single root invariant"](https://github.com/vito/progrock/pull/11))

vito commented on cmd/dagger/check.go:
```diff
@@ -0,0 +1,227 @@
+	var eg errgroup.Group
+	for _, check := range selectedChecks {
+		check := check
+		runCheckHierarchy(ctx, c, rec, path, &eg, check)
```
nit: missing an `err` check here

vito commented on cmd/dagger/check.go:
```diff
@@ -0,0 +1,227 @@
+	path = append([]string{}, path...)
+	path = append(path, name)
+	fullPathName := strings.Join(path, "->")
+	digest := digest.FromString(fullPathName)
```
It's tempting to use the check ID here. Or can the same check ID be present in multiple places in one `checks` call?

sipsma replied:
Yeah, and using the full path name is probably not needed after your fixes; that whole approach was just part of my original naive attempt at getting the progress output hierarchy working.

`CheckID` includes the environment ID (to enable invoking the check), so I think it should be safe to use... Will double check and update if so

vito commented on cmd/dagger/codegen.go:
```diff
@@ -0,0 +1,196 @@
+	}
+
+	if output == "" || output == "-" {
+		cmd.Println(string(generated))
```
super nit: maybe consider an early return here, for a moment I thought we were doing an `os.WriteFile("", ...)` down below, before analyzing the indentation

sipsma commented:
@vito thanks for the reviews so far! FYI I just pushed a commit that I'm almost done with. Just want to make sure that you aren't reviewing stale code (so far the comments are all applicable on a quick skim). I still need to fix up some failing tests in that most recent commit where dependency envs aren't always loaded when they need to be, but otherwise things seem to be working, so still safe to continue reviewing.

Will get back on the comments once I finish that up!

vito commented on core/schema/env.graphqls:
```diff
@@ -0,0 +1,252 @@
+
+  The entrypoint does not return any value.
+
+  In this case, the check's result is based on whether the entrypoint exit code.
```
```suggestion
  In this case, the check's result is based on the entrypoint exit code.
```

vito commented on core/schema/env.graphqls:
```diff
@@ -0,0 +1,252 @@
+
+  Exiting with any other exit code will be treated an internal error and not
+  result in the entrypoint execution being cached. This error will be forwarded
+  to any clients querying the check result.
```
Checking my understanding: we're able to be really restrictive here because the other side will always be implemented by one of our runtimes who uphold their end of this contract, rather than an exit code produced by running tests.

sipsma replied:
Yep exactly, for a little more background, the constraints we're working with here are:
1. We want SDKs to support multiple return types for a given entrypoint 
2. We want SDKs to be as dumb as possible and implement as much server-side as possible
3. For whatever SDKs do need to do, it's best if it's encoded in the api; or at min doc'd as part of that api
4. For checks, we need to distinguish between "internal errors" (which should not result in execution being cached) and "check errors" (which should result in the error being cached)

So the schema you commented on here is basically where I landed in terms of the engine<->sdk contract. SDKs need to tell the server what the specific entrypoint is returning (`Check`, `CheckResult`, `string`, void) and use different exit codes to distinguish between "internal error" and "user code error".
* I had a few different approaches to this earlier, but of the iterations that actually worked, this one resulted in the simplest code for the Go SDK

After that, the server does the rest of the magic needed to get the caching right. As part of this I actually added an internal ability for `Container.WithExec` to treat cache execution of execs with non-zero exit codes: https://github.com/sipsma/dagger/blob/179525ff35a1e4c5b338472e47d350d4dbee8fb2/core/container.go#L1855-L1855

Which could hopefully come in use later if we want to re-use that as part of our public api.

vito commented on core/schema/env.graphqls:
```diff
@@ -0,0 +1,252 @@
+  If set, sourceDirectorySubpath should point to the subpath of sourceDirectory that
+  contains the environment code. If unset, it will default to the root of the sourceDirectory.
+  """
+  from(
```
nit: `Container.From` gets a pass from me because it's so common and maps well to Dockerfile patterns, but this `from` feels like cheating. :joy: 

On the other hand, there's no real precedent for this. Everything else has a valid "zero" state which lets us build everything incrementally like `withName` and `withSDK`. Maybe this is just the easiest default name when there are multiple required values? Seems like the lesser evil to partial instantiation. The alternatives would be `new` or `create` but we don't want it to sound imperative, or that you are creating multiple instances when given the same args. Hmm...

sipsma replied:
Haha yeah you pretty much outlined the entire saga I went through here.

I started out trying to refactor the env schema to use `with*` on everything, but quickly realized it gets extremely confusing (both for users and us) that gets.

So I then pivoted to only have `with*` on the fields which make sense to change after initialization.

And once I reached that point, my brain was consumed 100% with refactoring code, so I went with `from` because I didn't have time/neurons to think about it more 😄

But totally agree this name is worth re-thinking. It makes sense on some level in that you are creating an environment *from* those parameters, but I don't know how obvious that is.

... I'm kind of drawing a blank here still 🤔. I'll keep thinking though.

vito commented on cmd/dagger/env.go:
```diff
@@ -0,0 +1,645 @@
+	environmentURI := environmentURI
+	if environmentURI == "" || environmentURI == environmentURIDefault {
+		// it's unset or default value, use env if present
+		if v, ok := os.LookupEnv("DAGGER_PROJECT"); ok {
```
Should probably be renamed:

```suggestion
		if v, ok := os.LookupEnv("DAGGER_ENV"); ok {
```

(feel free to ignore pending our open terminology questions though)

vito commented on codegen/generator/go/generator.go:
```diff
@@ -45,8 +66,28 @@ func (g *GoGenerator) Generate(_ context.Context, schema *introspection.Schema)
+
+			// don't create methods on query for the env itself, only its deps
+			// e.g. don't create `func (r *DAG) Go() *Go` in the Go env's codegen
+			if g.Config.EnvironmentName != "" && t.Name == generator.QueryStructName {
```
Interesting, so without this check would user envs have all of their entrypoints reflected on their own `dag` var after codegen?

Makes sense, but devil's advocate: theoretically it could be useful to call yourself through the DAG to leverage the entrypoint-level caching, instead of directly calling a sibling function. Would that "just work?" I'm not sure if it's worth exposing anyway, since it's a bit :exploding_head:. So far I've been able to write mostly pure entrypoints that just chain API calls and don't return `error`, so in that case the entrypoint-level caching would be pure overhead.

BTW: is there anything stopping someone from creating an env called `Container` and overriding the internal API?

sipsma replied:
> Interesting, so without this check would user envs have all of their entrypoints reflected on their own dag var after codegen?

Sort of, we actually stop that from happening in most cases much earlier than this code by only loading the env deps into the schema when generating for a given env.

I think this branch of code is leftover from the Zenith branch where we had universe, which was special because it was intrinsically loaded along side `core`. I'll try deleting this code from this branch to confirm.

---

Either way, yes totally agree being able to self call would be nice. The problem is the catch-22s you get yourself into when you're in the middle of writing env code, need to re-run codegen, but then it fails because your code doesn't compile. We ran into that in cloak and it was pretty maddening.

I would like to have this ability someday, but I think we'll need a radically different approach. My idea is that since we're already doing codegen + ast parsing, we can literally just take the user code and wrap it up such that when it's actually compiled, any direct call to another env entrypoint becomes wrapped in a call that goes through the actual server api (and thus wrapped in caching).

I'm pretty sure that isn't *quite* as crazy as it sounds; I think there's approaches to it that might actually be fairly robust. But it's still *sort of* crazy and thus something I'm not pursuing for now 🙂

---

> BTW: is there anything stopping someone from creating an env called Container and overriding the internal API?

Yes, that should result in an error because there will be a conflict during the schema merging. Sounds like a good test case though to be sure :-)

helderco replied:
> theoretically it could be useful to call yourself through the DAG to leverage the entrypoint-level caching, instead of directly calling a sibling function.

If I understand you correctly I plan on supporting this in Python, without needing or leveraging codegen:

https://github.com/dagger/dagger/blob/431ae2f5614596cb0a1ea779abee33319ffd07f7/sdk/python/src/dagger/server/_resolver.py#L271

General idea is this:

```python
from dagger.ext import function

@function
async def build():
    ...

@function
async def test():
    await build()  # calls function directly
    # or
    await build.call()  # calls function via API/DAG
    ...
```

This is possible because the decorator can return an object that wraps the function and it already has everything it needs to use the query builder underneath without relying on codegen.

There's precedent for this pattern. For example, in [Celery](https://docs.celeryq.dev/en/stable/userguide/calling.html#basics) you can do `my_task()` to call a function in the current process vs `my_task.apply_async()` to send it to a worker.

In [Prefect](https://docs.prefect.io/2.11.5/concepts/tasks/#tasks-overview) it's the opposite convention. You do `my_task.fn()` to call the function directly vs `my_task()` to run it in the executor. Also `my_task.submit()` for running async.

sipsma replied:
Makes total sense @helderco, this is definitely another area where I think we should be okay with each SDK having it's own sort of approach.

In Go, obviously nothing like that is really possible without adding overhead to the actual code the user writes (which I want to avoid at all costs, e.g. having them wrap `funcs` in some custom type that extends it with a method to call directly or with caching), so I think our only option there is to be more magical with some sort of codegen approach that rewrites funcs with that wrapping inherently.

helderco replied:
You're talking about always going through the API right? I was thinking on defaulting to a direct call since it seems less surprising, and explicitly requiring a `my_func.call()` to go through the API. That's the same as doing `dagger.call(my_func)`, which can be done in Go as well.

You think defaulting to go through the API makes the most sense (when making a direct call)? I would want to think so, but seems like more surprising to users. It's just a matter of documentation though (like with Prefect).

However it does make a weird case if I create a `@function` that uses only lazy API methods (not async), because going through the API forces it to be async so there's a signature mismatch that's easier to enforce with the added `my_task.call()` method, or external `call(my_task)`.

Should be the same in Go, requiring ctx and error.

sipsma commented on core/env.go:
```diff
@@ -0,0 +1,739 @@
+			if err != nil {
+				return nil, fmt.Errorf("failed to get stderr from check user container: %w", err)
+			}
+			// TODO: really need combined stdout/stderr now
```
Yep here's my thinking on this whole topic (broadly aligned with yours, and also a set of scattered loose thoughts 🙂):
1. It's important to our API that stdout/stderr are stored in the exec meta mount because that ties them to the cache of the exec.
   * Among other things, this ensures that our API works seamlessly with remote caching. E.g. `.Stdout()` is cached in the same way the filesystem of the exec is. 
   * This doesn't mean that we can't support streaming or anything. It's just an important implementation constraint to keep in mind for all of this.
2. We do indeed have the plumbing for streaming now, e.g. this method added as part of the previous re-arch mounts a file from buildkit's cache and streams it to the client (but for filesync): https://github.com/sipsma/dagger/blob/179525ff35a1e4c5b338472e47d350d4dbee8fb2/engine/buildkit/filesync.go#L254-L254
3. Whether we stream or return giant strings or both, there's either way some low level work needed to make combined stdout/stderr work as expected
   * Namely, in order to support all 3 APIs of `Stdout`, `Stderr` and `CombinedOutput`, we need to store the two streams of data in such a way that they can be presented both separately and together, and when combined it would preferably be interleaved as close as possible to the same order they were written to the pipes.
   * My thought has been that we might need to continue to use separate pipes for stdout/stderr, but then as we read lines from them, store the lines structured in some way that indicates which stream they were from (e.g. with some "header" like `[stdout]`/`[stderr]`) and then write them all to the same file with that structure.
   * That way, if you want just the stdout, you read the file, grab the stdout lines, strip the structure and return that stream. Same for stderr. And for combined, just read all the lines.
   * Probably other approaches too though.
4. I think you're spot on about the fact that being limited to "stdout/stderr of the most recent exec" being very limited.
   * While this would be tricky in the details, I can definitely see it making sense to be able to walk the DAG underlying a container and grab all the stdout/stderr/combined streams of the focused execs only. It's just a matter of figuring out stuff like whether we only grab the "most recent series of focused execs", "all focused execs", etc.
   * I also wouldn't want this to be the only API; it should hopefully just supplement the ability to get this data from an individual exec.

If you're in agreement about `3`, I think that we could at least implement that part as the foundation for everything else (whenever there's time obviously, it isn't exactly an emergency).

sipsma commented:
Cleaned up the last egregious hack (some global vars in the go codegen, did some refactoring of the code there to make template funcs parameterizable rather than a global map, which allowed the other global vars I added to also go away).

So bringing this out of Draft. Still comments to address (thank you @vito!), linters to appease, and some more test coverage I'd like to add (though there's a halfway decent set of integ tests already), but pretty close now. 

Also, many of the checks are failing due to 500s from dockerhub atm, so need to sift through what's real there :-)

sipsma commented on codegen/generator/nodejs/templates/functions.go:
```diff
@@ -41,7 +41,6 @@ var (
 		"ToLowerCase":         commonFunc.ToLowerCase,
 		"ToUpperCase":         commonFunc.ToUpperCase,
 		"ToSingleType":        toSingleType,
-		"GetEnumValues":       getEnumValues,
```
Not intentional at all... I don't think I ever even changed any files under `codegen/generator/nodejs` so confused how that would even have happened in a rebase, but good catch! Will fix

sipsma commented on core/schema/env.graphqls:
```diff
@@ -0,0 +1,252 @@
+extend type Query {
```
@shykes this is the file to bikeshed :-)

vito commented on core/integration/container_test.go:
```diff
@@ -2606,11 +2606,6 @@ func TestContainerExport(t *testing.T) {
 		require.NoError(t, err)
 		require.True(t, ok)
 
-		stat, err := os.Stat(imagePath)
-		require.NoError(t, err)
-		require.NotZero(t, stat.Size())
-		require.EqualValues(t, 0o600, stat.Mode().Perm())
-
```
Looks like some of the changes from https://github.com/dagger/dagger/pull/5649 got dropped too

sipsma replied:
wtf... I literally just `git pull -r upstream main`'d... maybe I'll just `git reflog` my way back and try again 🤷‍♂️

sipsma replied:
Thank you for catching all this though!

vito commented on core/integration/engine_test.go:
```diff
@@ -133,117 +130,3 @@ func TestEngineSetsNameFromEnv(t *testing.T) {
 	require.Contains(t, out, "Connected to engine "+engineName)
 }
-
-func TestDaggerRun(t *testing.T) {
```
These two tests, too :detective:

vito commented on core/integration/pipeline_test.go:
```diff
@@ -108,36 +89,3 @@ func TestPipeline(t *testing.T) {
 	})
 }
-
-func TestInternalVertexes(t *testing.T) {
```
Here too

sipsma commented on ci/main.go:
```diff
@@ -1,42 +1,367 @@
-// Dagger SDK targets
-func (t CITargets) SDK(ctx dagger.Context) (SDKTargets, error) {
-	return SDKTargets(t), nil
+func cniPlugins() *Directory {
```
backporting the recent changes on main to build this stuff from source is on my TODO list

sipsma commented on codegen/generator/functions.go:
```diff
@@ -102,12 +103,21 @@ func (c *CommonFunctions) GetArrayField(f *introspection.Field) []*introspection
 		if typeField.TypeRef.IsScalar() {
 			fields = append(fields, typeField)
 		}
+		// TODO: hack to fix requesting all fields from list of id-able objects, need better solution
```
reminder to self: double check this (I think it may actually be totally fine) and spin out to separate PR against main if so

sipsma commented on cmd/dagger/check.go:
Good catch, there was a ton of convoluted cobra stuff leftover from the `dagger do` implementation where we have to  handle the fact that there's `dagger` cmds/flags and user cmds/flags, which resulted in this. I just removed all that for now since it's not relevant for checks and I'd rather retry a new cleaner implementation of all that once we are merging commands.

Fixed it, but will also add an integ test for `--help` on all the subcommands now too.

sipsma replied:
Added integ tests here: https://github.com/dagger/dagger/pull/5659/commits/95f8639b9cc7247bb03c7527140f547e804453a9

sipsma commented on cmd/shim/main.go:
```diff
@@ -408,9 +434,10 @@ func setupBundle() int {
+				return errorExitCode
 			}
 			keepEnv = append(keepEnv, "_DAGGER_SERVER_ID="+execMetadata.ServerID)
+			keepEnv = append(keepEnv, "_DAGGER_ENVIRONMENT_DIGEST="+execMetadata.EnvironmentDigest.String())
```
TBH I hate the complication this added to the code (`CurrentEnvironment` is inherently complicated since it is self-referential and thus creates circular dependencies unless we are very careful), but yeah, it's neat now that it works! I'm also hopeful that we can find some simpler patterns for this in a future iteration, it felt like something that could be improved with another round of thought/eyes.

sipsma commented:
@vito I just rebase on main and picked up the interactive TUI changes and made the change you suggested to delete these lines [here](https://github.com/dagger/dagger/pull/5659/files#r1309335581).

When I run `dagger checks -e ci --focus=false` (I just like the unfocused output for those, it looks cool 😎), the output itself looks good but I can no longer exit the command (either `ctrl-C` or `q`). Did I miss something when updating?

vito commented:
Pausing for now, I was able to confirm that the `ctx` passed along is indeed canceled at the right time, but I didn't find which spot is ignoring it.

helderco commented on core/schema/env.graphqls:
```diff
@@ -0,0 +1,252 @@
+  """
+  Internal-only.
+
+  The name of the entrypoint to invoke. If not set, then the sdk
+  is expected to return the environment definition.
```
How about adding an `Entrypoint.sync` field to load the `with_check` calls, etc, instead of via `returnEntrypointValue`?

Pseudo code (serve):
```python
env = dagger.current_environment()
input_ = env.entrypoint_input()
name = await input_.name()

if not name:
    await env.sync()
    return

...
```

sipsma replied:
I actually started there; there were a few tricky parts because when we `Sync` this object in particular, the ID actually changes, so we hit the case you described [here](https://github.com/sipsma/dagger/blob/21e8e442736b0bbd301a8aca8e4b81e36fc309ad/codegen/generator/functions.go#L134-L142).

Those would be solvable with extra effort, but I ultimately went with the `returnEntrypointValue` approach anyways since not only avoids that whole problem but also ends up minimizing SDK responsibilities, moves as much to the engine as possible and gives us long term flexibility to modify how this works without updating every SDK:
1. It's important for the implementation that the result we return is actually written to filesystem of the entrypoint exec (so that the result is cached as part of that exec).
   * Previously, that meant every SDK needed to know where to write that file. Now, they are agnostic to it; they just need to call that API.
2. Centralizing what we write there entirely server-side is nice since we can very easily update it as needed, without any considerations around whether we'd impact the (non-internal) API of `.Sync`. It's highly plausible we'll want to make updates to the format of what we write or the details in how we process it since this is all intertwined with how the caching works.

helderco replied:
Yeah I understand but it can be any other name like `Environment.load` that's marked internal. Otherwise this looks a bit odd:

```python
await env.environment_result_value(str(await env.id()))
```

You have to serialize/marshal in the SDK, etc, would be simpler to just:

```python
await env.load()
```

I was confused if I can use the same environment instance because go sends the result value from a clean `dag.CurrentEnvironment()`. Is this required?

wingyplus commented:
I guess that Elixir failure because codegen handle incorrectly. Maybe it fixed with codegen improvement in  https://github.com/dagger/dagger/pull/5557 plus some addtional change. I’ll looking a bit deeper tomorrow. :)

sipsma commented:
Meant to close this before vacation, new iteration is here: https://github.com/dagger/dagger/pull/5757
